library/Network/CloudSeeder/Main.hs

```diff
@@ -98,53 +125,131 @@ runAppM (AppM x) = do
+collectParameters config stackToDeploy envVars env passedParams outputs =
+  let globalParams = config ^. parameters
       localParams = stackToDeploy ^. parameters
-      allParams = globalParams ++ localParams
-      isRequired = (`elem` requiredParams)
-      stackParams = filter (isRequired . fst) (allParams ++ outputs ++ envVars)
+  in globalParams <> localParams <> outputs <> envVars <> [("Env", env)] <> passedParams
 
-  csId <- computeChangeset (mkStackName nameToDeploy) templateBody stackParams stackTags
-  runChangeSet csId
+getParameters :: (MonadArguments CliError f, HasEnvironmentVariables s [T.Text], HasEnvironmentVariables s1 [T.Text], MonadEnvironment f, MonadCloud f, Traversable t, HasParameterSources s1 (S.Set (T.Text, ParameterSource)), HasParameterSources s (S.Set (T.Text, ParameterSource)), HasParameters s1 (S.Set (T.Text, T.Text)), HasParameters s (S.Set (T.Text, T.Text))) 
+              => s1 -> s -> t T.Text -> T.Text -> T.Text -> f (S.Set (T.Text, T.Text))
+getParameters config stackToDeploy dependencies env appName = do 
+  envVars <- getEnvVars config stackToDeploy
+  outputs <- getOutputs dependencies env appName
+  passedParams <- getPassedParameters config stackToDeploy
+  return $ collectParameters config stackToDeploy envVars env passedParams outputs
+
+validateTags :: (MonadError e m, AsCliError e) => S.Set (T.Text, T.Text) -> m (M.Map T.Text T.Text)
+validateTags ts = assertUnique _CliDuplicateTagValues ts
+
+validateParameters :: (AsCliError e, HasParameterSpecs s (f ParameterSpec), Foldable f, MonadError e m) 
+                   => s -> S.Set (T.Text, T.Text) -> m (M.Map T.Text T.Text)
+validateParameters template params = do
+  let paramSpecs = template ^. parameterSpecs
+      requiredParamNames = S.fromList (paramSpecs ^.. folded._Required)
+      allowedParamNames = S.fromList (paramSpecs ^.. folded.parameterKey)
+
+  uniqueParams <- assertUnique _CliDuplicateParameterValues params
+  let missingParamNames = requiredParamNames S.\\ M.keysSet uniqueParams
+  unless (S.null missingParamNames) $
+    throwing _CliMissingRequiredParameters missingParamNames
+
+  return $ uniqueParams `M.intersection` M.fromSet (const ()) allowedParamNames
 
 cliIO :: AppM DeploymentConfiguration -> IO ()
 cliIO mConfig = runAppM $ cli mConfig
 
+tuplesToMap :: Ord a => [(a, b)] -> M.Map a [b]
+tuplesToMap xs = M.fromList $ map concatGroup grouped
+  where
+    grouped = groupBy ((==) `on` fst) xs
+    concatGroup ys = (fst (head ys), map snd ys)
+
+assertUnique :: (AsCliError e, MonadError e m) => Prism' e (M.Map T.Text [T.Text]) -> S.Set (T.Text, T.Text) -> m (M.Map T.Text T.Text)
+assertUnique _Err paramSet = case duplicateParams of
+    [] -> return $ M.fromList paramList
+    _ -> throwing _Err duplicateParams
+  where
+    paramList = S.toList paramSet
+    paramsGrouped = tuplesToMap paramList
```

Comment:
I realized this code is not actually 100% correct, because `groupBy` only groups *contiguous* elements of a list. We can fix that by ensuring `paramList` is sorted, so we should use `S.toAscList` here to be safe.